Dead time models + Fourier Amplitude Difference correction

matteobachetti opened the pull request:
In this PR:
+ Non-paralyzable dead time model by Zhang+95
+ Dead time _filters_ to mimick both paralyzable and non-paralyzable dead time in simulated (or even real) data.
+ Aaaand another long-standing desideratum, the FAD. The current implementation does not make much use of the existing `Powerspectrum` etc. classes, because I fear that the code might become too convoluted. I chose, in this case, to just return a table containing the Leahy-normalized and FAD-corrected PDS from both light curves, the one from the summed light curves, and the cospectrum.

Resolve #106
Resolve #355 
Close #249

pep8speaks commented:
Hello @matteobachetti! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`stingray/deadtime/fad.py`](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/fad.py):

> [Line 23:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/fad.py#L23): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 28:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/fad.py#L28): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 30:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/fad.py#L30): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 32:25](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/fad.py#L32): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\&'
> [Line 163:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/fad.py#L163): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)



* In the file [`stingray/deadtime/tests/test_fad.py`](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py):

> [Line 10:1](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L10): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 18:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L18): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 79 characters)
> [Line 31:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L31): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 32:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L32): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 38:27](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L38): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 39:27](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L39): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 40:27](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L40): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 41:27](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L41): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 54:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L54): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 60:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L60): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 61:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L61): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 62:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L62): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 63:80](https://github.com/StingraySoftware/stingray/blob/2b3a32a2dcac7b931db11480e77a997337c5ebf0/stingray/deadtime/tests/test_fad.py#L63): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)

matteobachetti commented:
Oh, weird. Tests for the FAD are not running

dhuppenkothen commented on stingray/deadtime/filters.py:
```diff
@@ -0,0 +1,161 @@
+        del order
+    else:
+        tot_ev_list = ev_list
+        ev_kind = np.ones(len(ev_list), dtype=bool)
```
I think for future potential work in terms of changes/debugging, it would be really helpful to have some comments in this function. It gets a bit confusing with the different event lists and `ev_kind` and `all_ev_kind` that having a few comments in there to talk through the logic would be super helpful.

dhuppenkothen commented on stingray/deadtime/filters.py:
```diff
@@ -0,0 +1,161 @@
+    ----------------
+    bkg_ev_list : array-like
+        A background event list that affects dead time
+    dt_sigma : float
```
I think it's worth being specific here that this will result in dead time being draws from a normal distribution with `mean = deadtime` and `variance = dt_sigma`

matteobachetti replied:
Done

dhuppenkothen commented on stingray/deadtime/filters.py:
```diff
@@ -0,0 +1,161 @@
+    if isinstance(event_list, EventList) and hasattr(event_list, 'pi') and event_list.pi is not None:
+        warnings.warn(
+            "PI information is lost during dead time filtering",
+            AstropyUserWarning)
```
Why are we losing PI information during dead time filtering? Aren't we keeping track of which events are kept and which ones are lost, anyway, that we could apply that mask to the PI channels? Or am I missing something?

matteobachetti replied:
Ok. (Just being lazy :P). Done

dhuppenkothen commented on stingray/deadtime/model.py:
```diff
@@ -0,0 +1,151 @@
+
+
+@njit()
+def factorial(n):
```
Is this faster than the scipy implementation? If so, I think we might want to use this in other parts of the code (e.g. the modelling subpackage) where these functions appear ...

matteobachetti replied:
It's just a wrapper around the numba-compiled Gamma function. I don't know if it's faster frankly :)

dhuppenkothen replied:
Is there an easy way to test this? I don't think I have numba and I've never used numba before, but if it is, I think this would be super useful to have in `utils.py` rather than here, because there are parts of the modelling that heavily rely on factorials and Gamma functions

matteobachetti replied:
Ok, I made some quick test and scipy's version is faster (tens of microseconds instead of milliseconds) and more reliable (this version spits `inf` with factorial(1000), scipy's gives a result..).

matteobachetti replied:
Aaaah NOW I remember why. Numba is very picky when it comes to the functions you can call from inside numba functions. It did not accept Scipy's function. 
I changed again. Now I just create a lookup table with the first 150 factorials, calculated with Scipy, at import time, and numba is able to use that. This is still faster, because there are no repeated calculations.

dhuppenkothen commented on stingray/deadtime/model.py:
```diff
@@ -0,0 +1,151 @@
+
+
+@njit()
+def h(k, n, td, tb, tau):
```
What does this function do? The function name isn't very descriptive, so adding a docstring might help? Also, where is it used?

dhuppenkothen commented on stingray/deadtime/model.py:
```diff
@@ -0,0 +1,151 @@
+
+
+@njit()
+def A(k, r0, td, tb, tau):
```
Same as above. Yay for docstrings? :)

dhuppenkothen commented on stingray/deadtime/model.py:
```diff
@@ -0,0 +1,151 @@
+
+
+def check_A(rate, td, tb, max_k=100, save_to=None):
+    """Ak ->r0**2tb**2 for k->infty"""
```
???

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+from ..gti import cross_two_gtis, bin_intervals_from_gtis
+
+
+def _get_fourier_intv(lc, start_ind, end_ind):
```
Done

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+
+    Parameters
+    ----------
+    lc1: class:`stingray.ligthtcurve.Lightcurve`
```
The light curves need not be _identical_, but surely the flux must be similar. Also, it cannot apply to light curves in different energy bins from the same instrument, because dead time would not be independent. I modified the docstring accordingly.

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+    Other parameters
+    ----------------
+    plot : bool
+        Plot diagnostics
```
Done

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+    ----------
+    lc1: class:`stingray.ligthtcurve.Lightcurve`
+    lc1: class:`stingray.ligthtcurve.Lightcurve`
+    segment_size: float
```
Ok. Done

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+        Smoothing algorithm
+    smoothing_length : int
+        Number of bins to smooth in gaussian window smoothing
+    verbose: bool
```
Yes, I agree. But there is no coherent mechanism yet afaik

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+        stdtheor = 2 / np.sqrt(n)
+        std = (average_corrected_fourier_diff / n).std()
+        np.abs((std - stdtheor) / stdtheor) < tolerance
+    strict : bool
```
Done

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+
+    """
+    if smoothing_length is None:
+        smoothing_length = segment_size * 3
```
Done

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+    fourier = scipy.fftpack.fft(counts)
+
+    freq = scipy.fftpack.fftfreq(len(time), lc.dt)
+    good = freq > 0
```
Having the negative frequencies, if I'm not mistaken, is useful when we are interested in the complex quantities. Otherwise, the PDS and the other real-defined periodograms are just the same in the positive and negative frequencies. Am I missing something?

dhuppenkothen replied:
I don't know, though I'm sure you're right. @abigailStev is the person to ask here. :) 
I just want to be sure we don't accidentally loose out on some functionality.

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+            plt.plot(freq, f1_leahy, zorder=5, lw=1)
+            plt.plot(freq, f2_leahy, zorder=5, lw=1)
+
+        ptot += pt
```
The length of the zeros depends on the fact that the number of light curve bins is even or odd (because of the selection of positive frequencies), so I think it would require an ugly piece of code with "ifs" that are bug-prone (what if numpy changes slightly the fft algorithm?) and not even needed for the final result. However, I added a comment to explain what will happen.

matteobachetti commented on stingray/deadtime/filters.py:
```diff
@@ -0,0 +1,161 @@
+        ev_list = event_list
+        mjdref = 0
+
+    if deadtime <= 0.:
```
Ok. Raising ValueError from now on.
If it's 0, it means we are very lucky. Might be that someone is running simulations where dead time goes from 0 to a certain value. If 0, just return the original event list.

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+    if strict:
+        assert is_compliant
+
+    results = Table()
```
So, I would like to save all results together as a single table, mostly to save space. I introduced a couple of new things:
1. The table with the results can be saved to a file. This might be useful in case we need to dump the results of a large calculation for future use.
2. I wrote a new `get_periodograms_from_FAD_results` function that can accept either the table of the FAD correction results or the output file where the results are saved, and return the corresponding Stingray objects (e.g. `get_periodograms_from_FAD_results(results, kind='ptot')` returns the PDS from the summed light curves, kind='cs' the cospectrum etc.).
Do you think it's a good compromise?

dhuppenkothen replied:
I get the space saving issue, but I also think that running basically *any* stingray function and getting back an Astropy Table would have me very confused, because that isn't really what we return in any other part of the code. Is dumping the results to file the most likely use case? Are we very worried about the size of periodograms that we need to avoid duplication? I'm not trying to be difficult, I genuinely would like to understand where the bottlenecks are. 
It might be valuable to have a boolean option on the FAD function `return_objects` that if true, will return the objects, otherwise an Astropy Table, and which can be set to `False` by default if we think that most users will want a table.

matteobachetti replied:
Ok, `return_objects` implemented and tested :)

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+               compl='NOT ' if not is_compliant else '',
+               additional='Maybe something is not right.' if not is_compliant else '')
+
+    print(verbose_string)
```
It was a leftover debugging thing. It's not supposed to be there unless the conditions right below that line are met

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+
+    Returns
+    -------
+    results : class:`astropy.Table` object
```
See below

matteobachetti commented:
@dhuppenkothen: thanks for the thorough review! There are a few more things to adjust, but I think most major stuff is done and for some other stuff I'm not sure how to proceed. I'll reply to some comments below:

> As I've written inline, I think it would be super useful to have the FAD model return PSD/CSD objects. I'm not sure I understand your reasoning that this would be too convoluted? I think instead of adding the astropy Table at the end, one could just instantiate an empty PSD and CSD object, and then populate it with the correct values (and deepcopy where necessary).

The FAD as it is implemented corrects all periodograms (single-channel, summed-channel PDS, cospectrum) at the same time and it saves them to a single table without repeating information (e.g. the frequency column). I added a function that takes this table and returns `AveragedXYZKJspectrum` objects with the correct values. 

> As implemented, it looks like the FAD only knows Leahy normalization? What normalization does it return if it's not Leahy? I think it would be useful if this were consistent with the normalizations in the PSD/CSD classes.

For now, it knows Leahy and none. I think we have the machinery to convert between normalizations inside the Averaged*spectrum classes, right? It's probably better to use that machinery (or improve it) instead of duplicating code to calculate all normalizations every time we create a periodogram.

> I'm also curious whether it would be possible to implement an option in the Crossspectrum class that, if True, automatically corrects the CSD for dead time (and optionally returns the corrected PSDs as well)?

The FAD requires to calculate the correction in each segment of the data, so we would need to basically replicate the code in `fad.py` inside the Crossspectrum class (in `_make_segment_spectra`). This is what I was referring to when I used the word "convoluted". 

> Same for the dead time models you've implemented, which don't seem to have any documentation attached to them? 

_that_ I hope I fixed :)

> How is the user supposed to apply for example the Zhang model? I actually think having an apply_deadtime_zhang method on Powerspectrum would be super useful.


As per the Zhang dead time model, what worries me is that it is seldom precise in correcting actual PDSs. The dead time is almost never as precise as required from the model. I would use it as a comparison tool rather than as a magic button that "corrects the PDS". For the brave, one can always divide the PDS by the model and multiply by 2 to get the Leahy normalization, but my feeling is that it will not be a clean correction.

> Same for the event list filtering: Would it be possible to implement a method apply_deadtime for EventList that does the filtering with minimal effort?

We use `EventList`s in the code in `filters.py`, so that would produce a circular import. I don't know how to deal cleanly with this situation.

matteobachetti commented on stingray/deadtime/fad.py:
```diff
@@ -0,0 +1,189 @@
+    average_diff = average_diff_uncorr = 0
+
+    if plot:
+        plt.figure()
```
Done

dhuppenkothen commented on stingray/deadtime/filters.py:
```diff
@@ -0,0 +1,237 @@
+import warnings
```
I just realized from your comment on the PR that we now have two modules called `filters.py`, one in the top-level `stingray` repo and here. I think this might be confusing? Could we add the code here simply to `events.py`? This would solve the confusion of having two modules of the same name and also the issue of circular imports. :)

matteobachetti replied:
Ok, I think I've done everything you suggest here, in a slightly different way to avoid circular imports and other complications.

matteobachetti commented:
@dhuppenkothen I think that, besides the freq > 0 thing, everything is implemented.
